cirq/ops/three_qubit_gates.py

```diff
@@ -145,6 +149,132 @@ def __str__(self) -> str:
+        if not isinstance(exponent, (int, float, sympy.Basic)):
+            return NotImplemented
+        return ThreeQubitDiagonalGate([
+            protocols.mul(angle, exponent, NotImplemented)
+            for angle in self._diag_angles_radians
+        ])
+
+    def _decompose_(self, qubits):
+        """An adjacency-respecting decomposition.
+
+        0: ───p_0───@──────────────@───────@──────────@──────────
+                    │              │       │          │
+        1: ───p_1───X───@───p_3────X───@───X──────@───X──────@───
+                        │              │          │          │
+        2: ───p_2───────X───p_4────────X───p_5────X───p_6────X───
+
+        where p_i = T**(4*x_i) and x_i solve the system of equations
+                    [0, 0, 1, 0, 1, 1, 1][x_0]   [r_1]
+                    [0, 1, 0, 1, 1, 0, 1][x_1]   [r_2]
+                    [0, 1, 1, 1, 0, 1, 0][x_2]   [r_3]
+                    [1, 0, 0, 1, 1, 1, 0][x_3] = [r_4]
+                    [1, 0, 1, 1, 0, 0, 1][x_4]   [r_5]
+                    [1, 1, 0, 0, 0, 1, 1][x_5]   [r_6]
+                    [1, 1, 1, 0, 1, 0, 0][x_6]   [r_7]
+        where r_i is self._diag_angles_radians[i].
+
+        The above system was created by equating the composition of the gates
+        in the circuit diagram to np.diag(self._diag_angles) (shifted by a
+        global phase of np.exp(-1j * self._diag_angles[0])).
+        """
+
+        a, b, c = qubits
+        if hasattr(b, 'is_adjacent'):
+            if not b.is_adjacent(a):
+                b, c = c, b
+            elif not b.is_adjacent(c):
+                a, b = b, a
+        sweep_abc = [common_gates.CNOT(a, b), common_gates.CNOT(b, c)]
+        phase_matrix_inverse = .25 * np.array(
+            [[-1, -1, -1, 1, 1, 1, 1], [-1, 1, 1, -1, -1, 1, 1],
+             [1, -1, 1, -1, 1, -1, 1], [-1, 1, 1, 1, 1, -1, -1],
+             [1, 1, -1, 1, -1, -1, 1], [1, -1, 1, 1, -1, 1, -1],
+             [1, 1, -1, -1, 1, 1, -1]])
+        shifted_angles_tail = [
+            angle - self._diag_angles_radians[0]
+            for angle in self._diag_angles_radians[1:]
+        ]
+        phase_solutions = phase_matrix_inverse.dot(shifted_angles_tail)
+        p_gates = [
+            pauli_gates.Z**(solution / np.pi) for solution in phase_solutions
+        ]
+
+        return [
+            p_gates[0](a),
+            p_gates[1](b),
+            p_gates[2](c),
+            sweep_abc,
+            p_gates[3](b),
+            p_gates[4](c),
+            sweep_abc,
+            p_gates[5](c),
+            sweep_abc,
+            p_gates[6](c),
+            sweep_abc,
+        ]
+
+    def _apply_unitary_(self, args: protocols.ApplyUnitaryArgs) -> np.ndarray:
+        if self._is_parameterized_():
+            return NotImplemented
+        for index, angle in enumerate(self._diag_angles_radians):
+            slice_index = '0b' + str(index % 2) + str(index % 4 // 2) + str(
```

Comment:
It looks like you're reversing the bit order? Instead of doing this, just directly pass in `index` and re-order the matrix rows or cols to make that work.

Reply:
Right, I was ensuring the endianness of `index` because I noticed, for example, `args.subspace_index(1) = (slice(None, None, None), 1, 0, 0, Elllipsis)` and 
`args.subspace_index(2) = (slice(None, None, None), 0, 1, 0, Elllipsis)` and so on.

I made it less messy in my latest commit, but I think directly passing `index` to `args.subspace_index` will still require some of this twiddling but let me know if you think I'm missing something.